File: contrib/pypy-compile-to-native/hexspeak.py
import sys
import re


def solve_recursive_count(words, currentLen, used, targetLength, cnt):
    for i in xrange(1, targetLength - currentLen + 1):
        for word in words[i]:
            if word in used:
                continue
            if i != targetLength - currentLen:
                solve_recursive_count(
                    words, currentLen + i, used + [word], targetLength, cnt)
            else:
                cnt[0] += 1


def solve_nonrecursive_count(words, targetLength):
    from collections import deque
    candidates = deque([([], 0)])
    cnt = 0
    while candidates:
        wordsSoFar, currentLen = candidates.popleft()
        if currentLen == targetLength:
            cnt += 1
        else:
            for i in xrange(1, targetLength - currentLen + 1):
                for word in words.get(i, []):
                    if word not in wordsSoFar:
                        candidates.append(
                            (wordsSoFar + [word], currentLen + i))
    return cnt


def main(argv):
    words = [[] for _ in xrange(128)]
    goodwords = [
        "abbe",
        "abed",
        "accede",
        "acceded",
        "ace",
        "aced",
        "add",
        "added",
        "baa",
        "baaed",
        "babe",
        "bad",
        "bade",
        "bead",
        "beaded",
        "bed",
        "bedded",
        "bee",
        "beef",
        "beefed",
        "cab",
        "cabbed",
        "cad",
        "cafe",
        "caff",
        "ceca",
        "cede",
        "ceded",
        "dab",
        "dabbed",
        "dace",
        "dad",
        "dded",
        "dead",
        "deaf",
        "deb",
        "decade",
        "decaf",
        "decaff",
        "deed",
        "deeded",
        "def",
        "deface",
        "defaced",
        "ebb",
        "ebbed",
        "eff",
        "efface",
        "effaced",
        "effed",
        "fab",
        "facade",
        "face",
        "faced",
        "fad",
        "fade",
        "faded",
        "faff",
        "faffed",
        "fed",
        "fee",
        "feed"
    ]
    targetLength = 14
    letters = 'abcdef'
    for word in goodwords:
        if word not in words[len(word)]:
            words[len(word)].append(word)
    words[1] = ['a']
    cnt = [0]
    import time
    start = time.time()
    solve_recursive_count(words, 0, [], targetLength, cnt)
    end = time.time()
    print cnt[0], "in", 1000*(end-start), "ms"
    return 0
    #
    # Not using recursion is much slower! Apparently cache is thrashed
    #
    # results = solve_nonrecursive_count(words, targetLength)
    # print "Total:", results
    # import resource # Proof this has much more memory impact...
    # print resource.getrusage(resource.RUSAGE_SELF).ru_maxrss, "KB"

def target(driver,args):
    return main,None

if __name__ == "__main__":
    main(sys.argv)

File: contrib/hexspeak.js
var g_words = {
    1: ['a'],
    2: [],
    3: ['ace', 'add', 'baa', 'bad', 'bed', 'bee', 'cab', 'cad', 'dab', 'dad', 'deb', 'def', 'ebb', 'eff', 'fab', 'fad', 'fed', 'fee'],
    4: ['abbe', 'abed', 'aced', 'babe', 'bade', 'bead', 'beef', 'cafe', 'caff', 'ceca', 'cede', 'dace', 'dded', 'dead', 'deaf', 'deed', 'face', 'fade', 'faff', 'feed'],
    5: ['added', 'baaed', 'ceded', 'decaf', 'ebbed', 'effed', 'faced', 'faded'],
    6: ['accede', 'beaded', 'bedded', 'beefed', 'cabbed', 'dabbed', 'decade', 'decaff', 'deeded', 'deface', 'efface', 'facade', 'faffed'],
    7: ['acceded', 'defaced', 'effaced']
}
var count = 0;

function solve_recursive_count(words, currentLen, used, targetLength) 
{
    for (var i=1; i<=targetLength-currentLen; i++) {
        var words_of_length_i = words[i] || [];
        for(var j=0; j<words_of_length_i.length; j++) {
            word = words_of_length_i[j];
            if (used.indexOf(word) === -1)
                if (i !== targetLength - currentLen) {
                    used.push(word);
                    solve_recursive_count(words, currentLen + i, used, targetLength)
                    used.pop();
                } else
                    count++;
        }
    }
}

for(var b=0; b<10; b++) {
    count = 0;
    var st = new Date().getTime();
    solve_recursive_count(g_words, 0, [], 14);
    var en = new Date().getTime();
    console.log(count + " in " + (en-st));
}
